BUG: Fix dot and inner type/value exception failures

jakirkham opened the pull request:
Partial fix: https://github.com/numpy/numpy/issues/6990

**TL;DR:** Bug fix to catch a possible second exception while building an exception message. Includes tests for `dot` and `inner` where this bug was first discovered.

**History:** Initially, this PR began as tests for the current behavior of `dot` and `inner` when they have some mismatched types that cannot be coerced into some common working type. However, this resulted in discovering a special class of bugs in C where an exception could occur during the creation of the exception message, but was not being handled. This bug has been carefully described with a list of instances in this bug report (  https://github.com/numpy/numpy/issues/6990 ). As a result, this contains a fix for one instance of this bug and tests to verify that this instance is fixed.

jakirkham committed:
TST: Ensure `dot` fails correctly if array types cannot be coerced into a common type.

jakirkham committed:
TST: Ensure `inner` fails correctly if array types cannot be coerced into a common type.

jakirkham commented:
Seeing failure on `USE_DEBUG=1` only here, as well.  Not sure what is going on. Maybe related to this ( http://bugs.python.org/issue20500 ). Though it is supposedly fixed.

jakirkham commented:
This definitely feels like an interpreter bug. I tried running this with `python-dbg` version 2.7.11 and did not encounter this error.

jakirkham commented:
For clarity, this is what I am seeing...

    python3-dbg: ../Objects/object.c:463: PyObject_Repr: Assertion `!PyErr_Occurred()' failed.
    ./tools/travis-test.sh: line 108:  5108 Aborted                 (core dumped) $PYTHON ../tools/test-installed-numpy.py

seberg commented:
It sounds a bit like an error set without an error return (i.e. a C-function does PyErr_SetString, but then does not return NULL). The only thing I can think of is that it wants a PyErr_Clear() before setting the new error, though I don't remember that this was necessary.

jakirkham commented:
What about returning `-1`? It seems there are some cases `-1` is returned instead.

seberg commented:
Same as NULL in that case, there are two calling conventions for python function, those functions just don't return a python object in general.

jakirkham commented:
So, I have looked, but nothing obviously sticks out to me. The only thing I can think of is `PyArray_DescrFromType` return `NULL` in some cases, but doesn't use `PyErr_SetString` in all cases.

jakirkham commented:
The same can be said for `PyArray_FromAny`.

jakirkham commented:
I added `PyErr_Clear` above `PyErr_SetString` and it made no difference.

jakirkham commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -2041,6 +2041,13 @@ def __numpy_ufunc__(self, ufunc, method, pos, inputs, **kwargs):
+        A = np.array((1,1), dtype='i,i')
+
+        assert_raises(ValueError, np.dot, c, A)
+        assert_raises(TypeError, np.dot, A, c)
```
So, I managed to get this built so I can debug with `python3-dbg` interactively and discovered this line will cause the strange exception. However, the line before does not cause the exception.

jakirkham commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4825,6 +4832,13 @@ def test_matmul_inplace():
+        A = np.array((1,1), dtype='i,i')
+
+        assert_raises(TypeError, np.inner, c, A)
+        assert_raises(TypeError, np.inner, A, c)
```
Same story as above. This line will cause the exception, as well.

charris replied:
Just to add that this error occurs with python 3.5 debug also.

jakirkham commented:
So, the error in the second case looks something like this.

    TypeError: Cannot cast array data from dtype([('f0', '<i4'), ('f1', '<i4')]) to dtype('float64') according to the rule 'safe'

Based on what @charris has put forward, my guess is trying to create the representations of these `dtypes` to fill into the exception is causing the problem. Do either of you know where this is happening?

jakirkham commented:
In short, things like this are the problem. Somehow the exception needs to be stored/cleared before doing this and then recreated afterwards. ( https://github.com/numpy/numpy/blob/97c35365beda55c6dead8c50df785eb857f843f0/numpy/core/src/multiarray/array_assign_scalar.c#L207 )

jakirkham commented:
I have opened a new issue ( https://github.com/numpy/numpy/issues/6990 ) related to this problem.

jakirkham committed:
BUG: Clear error before constructing error message using calls to `PyObject_Repr`. Also, do a better job of handling any errors raised while constructing the error message.

jakirkham commented:
So, I tried to add a bug fix, but I am seeing a segmentation fault. However, I don't see it the first time I test the `dot` or `inner` case of this form `assert_raises(TypeError, <func>, A, c)`, but only the second time. Does anything look obviously wrong to either of you with this bug fix?

jakirkham commented:
Tried running `gdb` and it points me to this line ( https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/convert_datatype.c#L1010 ).

jakirkham commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1926,14 +1926,35 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
     /* Raise an error if the casting rule isn't followed */
     if (!PyArray_CanCastArrayTo(arr, newtype, casting)) {
```
After some more debugging with `gdb` apparently the segmentation fault is occurring at this line.

jakirkham commented:
So, this is kind of interesting. If I run the code below, I don't encounter the segmentation fault, but I do see the error message change after the first call. Namely, it changes to the error message raised by `PyArray_MatrixProduct2` from the one that we have tried to fix here. The `ValueError` will continue to show up every time I run it. A similar thing happens with `inner`. However, if I try to reassign a value to `A` (not `B`) it will segment fault.

    >>> import numpy
    >>> import numpy as np
    >>> A = np.array((1,1), dtype='i,i')
    >>> B = np.array((1.,1.), dtype='f,f')
    >>> np.dot(A, B)
    TypeError: Cannot cast array data from dtype([('f0', '<i4'), ('f1', '<i4')]) to dtype('V8') according to the rule 'safe'
    >>> np.dot(A, B)
    ValueError: Cannot find a common data type.
    ...
    >>> A = np.array((1,1), dtype='i,i')

jakirkham commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1926,14 +1926,35 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
         errmsg = PyUString_FromString("Cannot cast array data from ");
-        PyUString_ConcatAndDel(&errmsg,
-                PyObject_Repr((PyObject *)PyArray_DESCR(arr)));
+        arr_descr = PyArray_DESCR(arr);
```
I think this may have been the cause of the segmentation faults. At the time, I didn't realize this returned a borrowed references. So, I was effectively deleting part of the array, but leaving the rest of the array around. As a result, we saw some weird behavior due to this.

jakirkham commented:
Segmentation fault appears to be fixed locally. I think this will check out.

jakirkham commented:
Alright, looks like it checks out. Please let me know if you guys see anything else here that needs changing.

jakirkham commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1926,14 +1926,35 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
+            return NULL;
+        }
+        arr_descr_repr = PyObject_Repr((PyObject *)arr_descr);
+        Py_DECREF(arr_descr);
```
I think this may have been the cause of the segmentation faults. At the time, I didn't realize this returned a borrowed references. So, I was effectively deleting part of the array, but leaving the rest of the array around. As a result, we saw some weird behavior due to this.

jakirkham commented:
Failures on AppVeyor ( https://github.com/numpy/numpy/issues/6991 ) had not previously occurred for this PR with exception of the segmentation fault issue (also happened on Travis), which has since been resolved. I believe these to be unrelated to the content of this PR.

jakirkham commented:
As AppVeyor is merging with master, which is broken, it is currently failing the tests there. So, I ran my own AppVeyor build without this merge (this is rebased on a commit on `master` before the bad commit) and it looks like it checks out ( https://ci.appveyor.com/project/jakirkham/numpy/build/1.0.2 ).

njsmith commented:
I'm starting to get really confused at what I'm looking at here -- the PR description says it's new tests, the discussion is about some gnarly C bug fixes... Maybe in the future it would work better to bundle up the test changes and code changes and big fixes into one combined PR so they stay together as a unit? Or maybe I just haven't been paying enough attention, that happens too sometimes :-)

jakirkham commented:
It is pretty easy to get lost here. This is what happened.

* Made this PR ( https://github.com/numpy/numpy/pull/6968 ).
* Added simple tests for exception handling.
* Tests got weird interpreter crash in debug mode.
* Broke out the tests out in the current PR to see if they worked.
* Surprisingly still got an interpreter crash in debug mode.
* Spent some time with @seberg and @charris trying to understand weird interpreter crash.
* Figured out that this is due to one of many bugs fitting this class ( https://github.com/numpy/numpy/issues/6990 ).
* Added a bug fix here.

Sorry this is super confusing. It was never my intent to confuse anyone like this. The bug discovered was really surprising and quite subtle.

jakirkham commented:
Updated the PR title and description to better explain what is going on here now.

@njsmith, is this making more sense yet? If not please just ask, this was surprisingly tricky to find. The fix ends up being straightforward.

jakirkham commented:
> Maybe in the future it would work better to bundle up the test changes and code changes and big fixes into one combined PR so they stay together as a unit? Or maybe I just haven't been paying enough attention, that happens too sometimes :-)

They are. It's ok. :)

jakirkham commented:
Thanks everyone for your help. In particular, thanks to @seberg and @charris for helping me locate and fix this bug.
